node/modules/storageminer_idxprov.go

```diff
@@ -82,6 +82,26 @@ func IndexProvider(cfg config.IndexProviderConfig) func(params IdxProv, marketHo
 		// If announcements to the network are enabled, then set options for the publisher.
 		var e *engine.Engine
 		if cfg.Enable {
+			// Join the indexer topic using the market's pubsub instance. Otherwise, the provider
+			// engine would create its own instance of pubsub down the line in go-legs, which has
+			// no validators by default.
+			t, err := ps.Join(topicName)
+			if err != nil {
+				llog.Errorw("Failed to join indexer topic", "err", err)
+				return nil, xerrors.Errorf("joining indexer topic %s: %w", topicName, err)
+			}
+
+			// Get the miner ID and set as extra gossip data.
+			// The extra data is required by the lotus-specific index-provider gossip message validators.
+			ma := address.Address(maddr)
+			opts = append(opts,
+				engine.WithTopic(t),
+				engine.WithExtraGossipData(ma.Bytes()),
+			)
+			if cfg.Announce.AnnounceOverHttp {
+				opts = append(opts, engine.WithDirectAnnounce(cfg.Announce.DirectAnnounceURLs...))
+			}
+
 			// Advertisements can be served over the data transfer protocol
 			// (on graphsync) or over HTTP
 			if cfg.HttpPublisher.Enabled {
```

Comment:
We probably want to check that `PublicHostName` is not empty.

Reply:
Great; add a test case to `ToHttpMultiaddr`?

Reply:
Good call, turns out `multiaddr.NewMultiaddr()` accepts an empty hostname 🤔 

So I added an explicit check in [test: add test case for empty announce address hostname](https://github.com/filecoin-project/boost/pull/1452/commits/331879f300a07156ee71d4da1c75d02a31b0791e)